### PR TITLE
Feat/1541 setup mssql test database

### PR DESCRIPTION
**Pull request by driatic**
https://dev.azure.com/mimeronline/ONECore/_boards/board/t/Dev/Issues/?workitem=1541

## Description 

The idea of using an in-memory-database for unit tests does not work an practice because of big limitations in SQLite3. I've researched alternative in-memory db engines but there does not seem to be any good fit for our stack other than SQLite3. The main flaw with SQLite3 is that it lacks support for keywords used in our queries and migrations. The old WIP pr for in-memory is this: https://github.com/Bostads-AB-Mimer/onecore-leasing/pull/78

My second idea was to use https://testcontainers.com/ to create a mssql-container for a test database when running the tests. This was also good in theory but the performance is just way to slow and it makes the test suite very unpredictable when it comes to running time. 

## Current solution 

The least bad idea that I can come up with is to use a dedicated mssql test database. There are several options on how we could host this but I think that this is a decision that needs to be taken by the team and not me alone. 

## Todo 

We need a dedicated database for our unit tests. Currently the tests will requires a `.env.test` file with credentials to this database but we could also create separate env vars for this and scrap the loading of the `.env.test` file. The database could be manually created and hosted in our K8-cluster or created on demand by a container task in our GitHub action (the latter is for sure better but more complex). 

Regardless on hosting, we need to configure our GitHub action to support the test database credentials. 

The test suite now runs sequentially to not result in deadlocking the database tests. This can be improved but later performance does not seem to be of any issue. On my local machine the performance difference between runs in sequence and parallel is not noticeable.

### CODE REVIEW

File: src/services/lease-service/adapters/listing-adapter.ts [hunk 1]
```diff
@@ -306,14 +303,26 @@ const applicationExists = async (contactCode: string, listingId: number) => {
...
+  // If result is not null/undefined and not an empty object, return true
+  return true
 }
 
 const getExpiredListings = async () => {
   const currentDate = new Date()
   const listings = await db('listing')
     .where('PublishedTo', '<', currentDate)
-    .andWhere('Status', '==', ListingStatus.Active)
+    .andWhere('Status', '=', ListingStatus.Active)
```
Comment (driatic) [inline src/services/lease-service/adapters/listing-adapter.ts:325]: '==' was invalid MSSQL syntax

File: src/common/config.ts [hunk 2]
```diff
@@ -1,7 +1,12 @@
 import configPackage from '@iteam/config'
 import dotenv from 'dotenv'
+import path from 'path'
 
-dotenv.config()
+if (process.env.NODE_ENV == 'test') {
+  dotenv.config({ path: path.join(__dirname, '../../.env.test') })
+} else {
+  dotenv.config()
+}
```
Comment (momentiris) [inline src/common/config.ts:9]: När vi uppdaterar node till >20.6.0 så kan vi skippa dotenv helt och hållet 👍

File: src/services/lease-service/adapters/db.ts [hunk 3]
```diff
@@ -1,7 +1,67 @@
...
+    },
+  }
+}
+
+const getConfigBasedOnEnvironment = () => {
+  const environment = process.env.NODE_ENV || 'dev'
+  return environment === 'test' ? getTestConfig() : getStandardConfig()
+}
+
+export const db = knex(getConfigBasedOnEnvironment())
```
Comment (momentiris) [inline src/services/lease-service/adapters/db.ts:38]: Känns som vi borde kunna undvika denna typen av logik och få till så att `Config.leasingDatabase` helt enkelt direkt har rätt värden beroende på env. Inte säker på hur man gör med configpaketet vi använder.

File: src/services/lease-service/adapters/listing-adapter.ts [hunk 4]
```diff
@@ -306,14 +302,26 @@ const applicationExists = async (contactCode: string, listingId: number) => {
       ListingId: listingId,
     })
     .first()
-  return !!result // Convert result to boolean: true if exists, false if not
+
+  // Check if result is null or undefined
+  if (!result) {
+    return false
+  }
```
Comment (momentiris) [inline src/services/lease-service/adapters/listing-adapter.ts:309]: När blir detta ett tomt objekt? 🤔

File: src/services/lease-service/tests/adapters/listing-adapter.test.ts [hunk 5]
```diff
@@ -11,83 +16,293 @@ jest.mock('onecore-utilities', () => {
...
-    expect(snd.applicants).toHaveLength(1)
-    expect(snd.applicants?.[0]?.listingId).toBe(snd.id)
+afterEach(async () => {
+  await db('applicant').del()
+  await db('listing').del()
+})
+
+afterAll(async () => {
+  await teardown()
+})
```
Comment (momentiris) [inline src/services/lease-service/tests/adapters/listing-adapter.test.ts:30]: migrate och teardown kan göras i den globala jest setup-filen `jest.config.js`

Comment (momentiris) [reply src/services/lease-service/tests/adapters/listing-adapter.test.ts:30]: Det hade varit gött om vi kunde få bort att manuellt behöva göra delete på alla tabeller som påverkas i en testfil. I ett annat projekt så wrappas alla tester som använder db i en transaktion och då försvinner behovet att rensa databasen då det görs en rollback efter varje test. Det blir svårt i vår kod just nu (tror jag) eftersom våra adapters importerar db-connection direkt från fil. 

Hade våra adapterfunktioner tagit db som parameter hade det varit enklare... 🤔

File: src/services/lease-service/adapters/listing-adapter.ts (hunk 4 above)
Comment (driatic) [reply src/services/lease-service/adapters/listing-adapter.ts:324]: Kan faktiskt inte svara på det längre och lyckas inte återskapa det nu heller 😇 Vet att jag null/undefined inte räckte tidigare men eftersom jag  nu inte kan återskapa det så tar jag bort den här checken.

**Pull request merged**